quality_of_service_demo/rclcpp/src/qos_overrides_listener.cpp

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/qos_overriding_options.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp_components/register_node_macro.hpp"
+
+#include "sensor_msgs/msg/image.hpp"
+
+#include "quality_of_service_demo/visibility_control.h"
+
+namespace quality_of_service_demo
+{
+class QosOverridesListener : public rclcpp::Node
+{
+public:
+  QUALITY_OF_SERVICE_DEMO_PUBLIC
+  explicit QosOverridesListener(const rclcpp::NodeOptions & options)
+  : Node("qos_overrides_listener", options)
+  {
+    // Callback that will be used when a message is received.
+    auto callback =
+      [this](const sensor_msgs::msg::Image::SharedPtr msg) -> void
+      {
+        rclcpp::Time now = this->get_clock()->now();
+        auto diff = now - msg->header.stamp;
+        RCLCPP_INFO(
+          this->get_logger(),
+          "I heard an image. Message single trip latency: [%f]",
+          diff.seconds());
+      };
+    rclcpp::SubscriptionOptions sub_opts;
+    // Update the subscription options to allow reconfigurable qos settings.
+    sub_opts.qos_overriding_options = rclcpp::QosOverridingOptions {
+      {
+        // Here all policies that are desired to be reconfigurable are listed.
+        rclcpp::QosPolicyKind::Depth,
+        rclcpp::QosPolicyKind::Durability,
+        rclcpp::QosPolicyKind::History,
+        rclcpp::QosPolicyKind::Reliability,
+      },
+      [](const rclcpp::QoS & qos) {
+        /** This is a qos validation callback, that can optionally be provided.
+         * Here, arbitrary constraints in the final qos profile can be checked.
+         * The function will return true if the user provided qos profile is accepted.
+         * If the profile is not accepted, the user will get an InvalidQosOverridesException.
+         */
+        // TODO(ivanpauno): Add getters to `rclcpp::QoS` to make this easier.
```

Comment:
I was planning to address this TODO after merging https://github.com/ros2/rclcpp/pull/1467 which is now ready (only waiting an approval) but I don't mind if this PR is merged first, I can address the TODO in a follow-up.

Reply:
Sounds good to wait. I don't think there's a rush to get this PR merged.

Reply:
Done in https://github.com/ros2/demos/pull/474/commits/7e57e571c36d455189861dfa0a698ceb577f1c23